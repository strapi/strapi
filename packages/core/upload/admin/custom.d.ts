export {};

declare global {
  interface Window {
    strapi: {
      backendURL: string;
    };
  }
  declare module '*?raw';
}
