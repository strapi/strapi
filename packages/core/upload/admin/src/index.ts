import { Images } from '@strapi/icons';

import pluginPkg from '../../package.json';

import { MediaLibraryDialog } from './components/MediaLibraryDialog/MediaLibraryDialog';
import { MediaLibraryInput } from './components/MediaLibraryInput/MediaLibraryInput';
import { PERMISSIONS } from './constants';
import { pluginId } from './pluginId';
import { getTrad, prefixPluginTranslations } from './utils';

import type { MediaLibraryDialogProps } from './components/MediaLibraryDialog/MediaLibraryDialog';
import type { MediaLibraryInputProps } from './components/MediaLibraryInput/MediaLibraryInput';
import type { StrapiApp } from '@strapi/admin/strapi-admin';
import type { Plugin } from '@strapi/types';

const name = pluginPkg.strapi.name;

const admin: Plugin.Config.AdminInput = {
  register(app: StrapiApp) {
    app.addMenuLink({
      to: `plugins/${pluginId}`,
      icon: Images,
      intlLabel: {
        id: `${pluginId}.plugin.name`,
        defaultMessage: 'Media Library',
      },
      permissions: PERMISSIONS.main,
      Component: () => import('./pages/App/App').then((mod) => ({ default: mod.Upload })),
      position: 4,
    });

    app.addSettingsLink('global', {
      id: 'media-library-settings',
      to: 'media-library',
      intlLabel: {
        id: getTrad('plugin.name'),
        defaultMessage: 'Media Library',
      },
      async Component() {
        const { ProtectedSettingsPage } = await import('./pages/SettingsPage/SettingsPage');
        return { default: ProtectedSettingsPage };
      },
      permissions: PERMISSIONS.settings,
    });

    app.addFields({
      type: 'media',
      Component: MediaLibraryInput as React.FC<Partial<MediaLibraryInputProps>>,
    });
    app.addComponents([
      {
        name: 'media-library',
        Component: MediaLibraryDialog as React.FC<Partial<MediaLibraryDialogProps>>,
      },
    ]);

    app.registerPlugin({
      id: pluginId,
      name,
    });
  },
  async registerTrads({ locales }: { locales: string[] }) {
    const importedTrads = await Promise.all(
      locales.map((locale) => {
        return import(`./translations/${locale}.json`)
          .then(({ default: data }) => {
            return {
              data: prefixPluginTranslations(data, pluginId),
              locale,
            };
          })
          .catch(() => {
            return {
              data: {},
              locale,
            };
          });
      })
    );

    return Promise.resolve(importedTrads);
  },
};

// eslint-disable-next-line import/no-default-export
export default admin;
