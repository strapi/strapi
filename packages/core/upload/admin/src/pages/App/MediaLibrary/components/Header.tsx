import { useQueryParams, Layouts } from '@strapi/admin/strapi-admin';
import { Button, Flex, Link } from '@strapi/design-system';
import { ArrowLeft, Plus } from '@strapi/icons';
import { stringify } from 'qs';
import { useIntl } from 'react-intl';
import { useLocation, NavLink } from 'react-router-dom';

import { Breadcrumbs } from '../../../../components/Breadcrumbs/Breadcrumbs';
import { getTrad } from '../../../../utils';

import type { Folder } from '../../../../../../shared/contracts/folders';
import type { CrumbDefinition } from '../../../../components/Breadcrumbs/Breadcrumbs';

interface FolderDefinition extends Omit<Folder, 'children' | 'files' | 'parent'> {
  children: {
    count: number;
  };
  files: {
    count: number;
  };
  parent?: FolderDefinition;
}

export interface HeaderProps {
  breadcrumbs?: Array<CrumbDefinition> | null;
  canCreate: boolean;
  folder?: FolderDefinition | null;
  onToggleEditFolderDialog: ({ created }?: { created?: boolean }) => void;
  onToggleUploadAssetDialog: () => void;
}

export const Header = ({
  breadcrumbs = null,
  canCreate,
  folder = null,
  onToggleEditFolderDialog,
  onToggleUploadAssetDialog,
}: HeaderProps) => {
  const { formatMessage } = useIntl();
  const { pathname } = useLocation();
  const [{ query }] = useQueryParams();
  const backQuery = {
    ...query,
    folder:
      folder?.parent && typeof folder.parent !== 'number' && folder.parent.id
        ? folder.parent.id
        : undefined,
    folderPath:
      folder?.parent && typeof folder.parent !== 'number' && folder.parent.path
        ? folder.parent.path
        : undefined,
  };

  return (
    <Layouts.Header
      title={formatMessage({
        id: getTrad('plugin.name'),
        defaultMessage: `Media Library`,
      })}
      subtitle={
        breadcrumbs &&
        typeof breadcrumbs !== 'boolean' &&
        folder && (
          <Breadcrumbs
            label={formatMessage({
              id: getTrad('header.breadcrumbs.nav.label'),
              defaultMessage: 'Folders navigation',
            })}
            breadcrumbs={breadcrumbs}
            currentFolderId={folder?.id}
          />
        )
      }
      navigationAction={
        folder && (
          <Link
            tag={NavLink}
            startIcon={<ArrowLeft />}
            to={`${pathname}?${stringify(backQuery, { encode: false })}`}
          >
            {formatMessage({
              id: getTrad('header.actions.folder-level-up'),
              defaultMessage: 'Back',
            })}
          </Link>
        )
      }
      primaryAction={
        canCreate && (
          <Flex gap={2}>
            <Button startIcon={<Plus />} variant="secondary" onClick={onToggleEditFolderDialog}>
              {formatMessage({
                id: getTrad('header.actions.add-folder'),
                defaultMessage: 'Add new folder',
              })}
            </Button>

            <Button startIcon={<Plus />} onClick={onToggleUploadAssetDialog}>
              {formatMessage({
                id: getTrad('header.actions.add-assets'),
                defaultMessage: 'Add new assets',
              })}
            </Button>
          </Flex>
        )
      }
    />
  );
};
