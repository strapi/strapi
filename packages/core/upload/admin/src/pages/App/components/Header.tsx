import { useQueryParams, Layouts } from '@strapi/admin/strapi-admin';
import { Button, Flex, Link } from '@strapi/design-system';
import { ArrowLeft, Plus } from '@strapi/icons';
import { stringify } from 'qs';
import { useIntl } from 'react-intl';
import { useLocation, NavLink } from 'react-router-dom';

import { Breadcrumbs } from '../../../components/Breadcrumbs/Breadcrumbs';
import { getTrad } from '../../../utils';

import type { Folder } from '../../../../../shared/contracts/folders';
import type { CrumbDefinition } from '../../../components/Breadcrumbs/Breadcrumbs';

interface HeaderProps {
  breadcrumbs?: Array<CrumbDefinition> | null;
  canCreate: boolean;
  folder?: Folder | null;
  onToggleEditFolderDialog: ({ created }?: { created?: boolean }) => void;
  onToggleUploadAssetDialog: () => void;
}

export const Header = ({
  breadcrumbs = null,
  canCreate,
  folder = null,
  onToggleEditFolderDialog,
  onToggleUploadAssetDialog,
}: HeaderProps) => {
  const { formatMessage } = useIntl();
  const { pathname } = useLocation();
  const [{ query }] = useQueryParams();
  const backQuery = {
    ...query,
    folder:
      folder?.parent && typeof folder.parent !== 'number' && folder.parent.id
        ? folder.parent.id
        : undefined,
  };

  return (
    <Layouts.Header
      title={formatMessage({
        id: getTrad('plugin.name'),
        defaultMessage: `Media Library`,
      })}
      subtitle={
        breadcrumbs &&
        folder && (
          <Breadcrumbs
            label={formatMessage({
              id: getTrad('header.breadcrumbs.nav.label'),
              defaultMessage: 'Folders navigation',
            })}
            breadcrumbs={breadcrumbs}
            currentFolderId={folder?.id}
          />
        )
      }
      navigationAction={
        folder && (
          <Link
            tag={NavLink}
            startIcon={<ArrowLeft />}
            to={`${pathname}?${stringify(backQuery, { encode: false })}`}
          >
            {formatMessage({
              id: getTrad('header.actions.folder-level-up'),
              defaultMessage: 'Back',
            })}
          </Link>
        )
      }
      primaryAction={
        canCreate && (
          <Flex gap={2}>
            <Button startIcon={<Plus />} variant="secondary" onClick={onToggleEditFolderDialog}>
              {formatMessage({
                id: getTrad('header.actions.add-folder'),
                defaultMessage: 'Add new folder',
              })}
            </Button>

            <Button startIcon={<Plus />} onClick={onToggleUploadAssetDialog}>
              {formatMessage({
                id: getTrad('header.actions.add-assets'),
                defaultMessage: 'Add new assets',
              })}
            </Button>
          </Flex>
        )
      }
    />
  );
};
