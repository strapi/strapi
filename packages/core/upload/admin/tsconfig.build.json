{
  "extends": "tsconfig/client.json",
  "compilerOptions": {
    "rootDir": "../",
    "baseUrl": ".",
    "outDir": "../dist",
    "noEmit": false
  },
  "include": ["./src", "./custom.d.ts", "../shared", "../package.json"],
  "exclude": ["**/__mocks__", "./src/**/tests", "**/*.test.*"]
}
