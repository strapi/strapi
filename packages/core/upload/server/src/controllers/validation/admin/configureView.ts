import { yup, validateYupSchema } from '@strapi/utils';
import { ALLOWED_SORT_STRINGS } from '../../../constants';

const configSchema = yup.object({
  pageSize: yup.number().required(),
  sort: yup.mixed().oneOf(ALLOWED_SORT_STRINGS),
});

export const validateViewConfiguration = validateYupSchema(configSchema);

export type ViewConfiguration = yup.InferType<typeof configSchema>;
