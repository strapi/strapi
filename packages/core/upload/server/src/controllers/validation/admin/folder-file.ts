import { intersection, map, isEmpty } from 'lodash/fp';
import { yup, validateYupSchema } from '@strapi/utils';
import { FOLDER_MODEL_UID } from '../../../constants';
import { folderExists } from './utils';
import { isFolderOrChild } from '../../utils/folders';

import type { Folder } from '../../../types';

const validateDeleteManyFoldersFilesSchema = yup
  .object()
  .shape({
    fileIds: yup.array().of(yup.strapiID().required()),
    folderIds: yup.array().of(yup.strapiID().required()),
  })
  .noUnknown()
  .required();

const validateStructureMoveManyFoldersFilesSchema = yup
  .object()
  .shape({
    destinationFolderId: yup
      .strapiID()
      .nullable()
      .defined()
      .test('folder-exists', 'destination folder does not exist', folderExists),
    fileIds: yup.array().of(yup.strapiID().required()),
    folderIds: yup.array().of(yup.strapiID().required()),
  })
  .noUnknown()
  .required();

const validateDuplicatesMoveManyFoldersFilesSchema = yup
  .object()
  .test('are-folders-unique', 'some folders already exist', async function areFoldersUnique(value) {
    const { folderIds, destinationFolderId } = value;
    if (isEmpty(folderIds)) return true;

    const folders = await strapi.db.query(FOLDER_MODEL_UID).findMany({
      select: ['name'],
      where: { id: { $in: folderIds } },
    });

    const existingFolders = await strapi.db.query(FOLDER_MODEL_UID).findMany({
      select: ['name'],
      where: { parent: { id: destinationFolderId } },
    });

    const duplicatedNames = intersection(map('name', folders), map('name', existingFolders));
    if (duplicatedNames.length > 0) {
      return this.createError({
        message: `some folders already exists: ${duplicatedNames.join(', ')}`,
      });
    }

    return true;
  });

const validateMoveFoldersNotInsideThemselvesSchema = yup
  .object()
  .test(
    'dont-move-inside-self',
    'folders cannot be moved inside themselves or one of its children',
    async function validateMoveFoldersNotInsideThemselves(value) {
      const { folderIds, destinationFolderId } = value;
      if (destinationFolderId === null || isEmpty(folderIds)) return true;

      const destinationFolder = await strapi.db.query(FOLDER_MODEL_UID).findOne({
        select: ['path'],
        where: { id: destinationFolderId },
      });

      const folders: Folder[] = await strapi.db.query(FOLDER_MODEL_UID).findMany({
        select: ['name', 'path'],
        where: { id: { $in: folderIds } },
      });

      const unmovableFoldersNames = folders
        .filter((folder) => isFolderOrChild(destinationFolder, folder))
        .map((f) => f.name);

      if (unmovableFoldersNames.length > 0) {
        return this.createError({
          message: `folders cannot be moved inside themselves or one of its children: ${unmovableFoldersNames.join(
            ', '
          )}`,
        });
      }

      return true;
    }
  );

export const validateDeleteManyFoldersFiles = validateYupSchema(
  validateDeleteManyFoldersFilesSchema
);

export async function validateMoveManyFoldersFiles(body: unknown) {
  await validateYupSchema(validateStructureMoveManyFoldersFilesSchema)(body);
  await validateYupSchema(validateDuplicatesMoveManyFoldersFilesSchema)(body);
  await validateYupSchema(validateMoveFoldersNotInsideThemselvesSchema)(body);
}
