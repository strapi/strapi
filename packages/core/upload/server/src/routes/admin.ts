export const routes = {
  type: 'admin',
  routes: [
    {
      method: 'GET',
      path: '/settings',
      handler: 'admin-settings.getSettings',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.settings.read'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/settings',
      handler: 'admin-settings.updateSettings',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.settings.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/',
      handler: 'admin-upload.upload',
      config: {
        policies: ['admin::isAuthenticatedAdmin'],
      },
    },
    {
      method: 'GET',
      path: '/files',
      handler: 'admin-file.find',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/files/:id',
      handler: 'admin-file.findOne',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'DELETE',
      path: '/files/:id',
      handler: 'admin-file.destroy',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folders/:id',
      handler: 'admin-folder.findOne',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folders',
      handler: 'admin-folder.find',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/folders',
      handler: 'admin-folder.create',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.create'],
            },
          },
        ],
      },
    },
    {
      method: 'PUT',
      path: '/folders/:id',
      handler: 'admin-folder.update',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'GET',
      path: '/folder-structure',
      handler: 'admin-folder.getStructure',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.read'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/actions/bulk-delete',
      handler: 'admin-folder-file.deleteMany',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
    {
      method: 'POST',
      path: '/actions/bulk-move',
      handler: 'admin-folder-file.moveMany',
      config: {
        policies: [
          'admin::isAuthenticatedAdmin',
          {
            name: 'admin::hasPermissions',
            config: {
              actions: ['plugin::upload.assets.update'],
            },
          },
        ],
      },
    },
  ],
};
