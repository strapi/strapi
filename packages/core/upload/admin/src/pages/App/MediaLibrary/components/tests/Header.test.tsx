import { useQueryParams } from '@strapi/admin/strapi-admin';
import { DesignSystemProvider } from '@strapi/design-system';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';

import { Header } from '../Header';

import type { HeaderProps } from '../Header';

jest.mock('@strapi/admin/strapi-admin', () => ({
  ...jest.requireActual('@strapi/admin/strapi-admin'),
  useQueryParams: jest.fn(() => [{ rawQuery: '', query: { folder: 2 } }, jest.fn()]),
}));

const FIXTURE_FOLDER = {
  id: 1,
  name: 'Folder 1',
  children: {
    count: 1,
  },
  createdAt: '',
  files: {
    count: 1,
  },
  path: '/1',
  pathId: 1,
  updatedAt: '',
  parent: {
    id: 2,
    name: 'Folder 2',
    children: {
      count: 1,
    },
    createdAt: '',
    files: {
      count: 1,
    },
    path: '/1',
    pathId: 1,
    updatedAt: '',
  },
};

const setup = (props: Partial<HeaderProps>) => {
  const withDefaults = {
    ...props,
    assetCount: 2,
    folderCount: 2,
    canCreate: true,
    onToggleEditFolderDialog: jest.fn(),
    onToggleUploadAssetDialog: jest.fn(),
  };

  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        refetchOnWindowFocus: false,
      },
    },
  });

  return render(
    <QueryClientProvider client={queryClient}>
      <DesignSystemProvider>
        <IntlProvider locale="en">
          <MemoryRouter>
            <Header {...withDefaults} />
          </MemoryRouter>
        </IntlProvider>
      </DesignSystemProvider>
    </QueryClientProvider>
  );
};

describe('Header', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders', () => {
    const { container } = setup({ folder: FIXTURE_FOLDER });

    expect(container).toMatchSnapshot();
  });

  test('does not render a back button at the root level of the media library', () => {
    const { queryByText } = setup({ folder: null });

    expect(queryByText('Back')).not.toBeInTheDocument();
  });

  test('does render a back button at a nested level of the media library', () => {
    (useQueryParams as jest.Mock).mockReturnValueOnce([
      { rawQuery: '', query: { folder: 2 } },
      jest.fn(),
    ]);
    const { getByText } = setup({ folder: FIXTURE_FOLDER });

    expect(getByText('Back')).toBeInTheDocument();
  });
});
