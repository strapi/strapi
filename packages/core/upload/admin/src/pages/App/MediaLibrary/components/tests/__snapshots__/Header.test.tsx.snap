// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders 1`] = `
.c0 {
  padding-block-start: 24px;
  padding-block-end: 40px;
  padding-inline-start: 56px;
  padding-inline-end: 56px;
  background: #f6f6f9;
}

.c1 {
  padding-block-end: 8px;
}

.c2 {
  cursor: pointer;
}

.c7 {
  min-width: 0;
}

.c11 {
  padding-inline-start: 16px;
  padding-inline-end: 16px;
  border-radius: 4px;
  cursor: pointer;
}

.c6 {
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
  display: flex;
}

.c8 {
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c10 {
  gap: 8px;
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c12 {
  gap: 8px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: inline-flex;
}

.c5 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: #4945ff;
}

.c9 {
  font-weight: 600;
  font-size: 3.2rem;
  line-height: 1.25;
  color: currentcolor;
}

.c14 {
  font-size: 1.2rem;
  line-height: 1.33;
  color: currentcolor;
  font-weight: 600;
}

.c16 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c3 {
  text-decoration: none;
}

.c3:visited {
  color: inherit;
}

.c13 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #d9d8ff;
  background: #f0f0ff;
  color: #271fe0;
}

.c13:hover {
  background-color: #ffffff;
}

.c13:active {
  background-color: #ffffff;
  border: 1px solid #4945ff;
  color: #4945ff;
}

.c13[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c15 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #4945ff;
  background: #4945ff;
  color: #ffffff;
}

.c15:hover {
  border: 1px solid #7b79ff;
  background: #7b79ff;
}

.c15:active {
  border: 1px solid #4945ff;
  background: #4945ff;
}

.c15[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c4 {
  display: inline-flex;
  align-items: center;
  text-decoration: none;
  gap: 8px;
  position: relative;
  outline: none;
}

.c4 svg {
  font-size: 1rem;
}

.c4 svg path {
  fill: #4945ff;
}

.c4:hover>span {
  color: #7b79ff;
}

.c4:hover svg path {
  fill: #7b79ff;
}

.c4:active {
  color: #271fe0;
}

.c4:after {
  transition-property: all;
  transition-duration: 0.2s;
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border: 2px solid transparent;
}

.c4:focus-visible {
  outline: none;
}

.c4:focus-visible:after {
  border-radius: 8px;
  content: '';
  position: absolute;
  top: -5px;
  bottom: -5px;
  left: -5px;
  right: -5px;
  border: 2px solid #4945ff;
}

@media (prefers-reduced-motion: no-preference) {
  .c13 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c15 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

<div>
  <div
    style="height: 0px;"
  >
    <div
      class="c0"
      data-strapi-header="true"
    >
      <div
        class="c1"
      >
        <a
          aria-current="page"
          aria-disabled="false"
          class="c2 c3 c4 active"
          href="/?folder=2&folderPath=/1"
        >
          <svg
            fill="currentColor"
            height="16"
            viewBox="0 0 32 32"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M28.5 16a1.5 1.5 0 0 1-1.5 1.5H8.625l6.44 6.439a1.502 1.502 0 1 1-2.125 2.125l-9-9a1.5 1.5 0 0 1 0-2.125l9-9a1.503 1.503 0 0 1 2.125 2.125L8.625 14.5H27a1.5 1.5 0 0 1 1.5 1.5"
            />
          </svg>
          <span
            class="c5"
          >
            Back
          </span>
        </a>
      </div>
      <div
        class="c6"
      >
        <div
          class="c7 c8"
        >
          <h1
            class="c9"
          >
            Media Library
          </h1>
        </div>
        <div
          class="c10"
        >
          <button
            aria-disabled="false"
            class="c11 c12 c13"
          >
            <span
              aria-hidden="true"
              class="c8"
            >
              <svg
                fill="currentColor"
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M28.5 16a1.5 1.5 0 0 1-1.5 1.5h-9.5V27a1.5 1.5 0 1 1-3 0v-9.5H5a1.5 1.5 0 1 1 0-3h9.5V5a1.5 1.5 0 1 1 3 0v9.5H27a1.5 1.5 0 0 1 1.5 1.5"
                />
              </svg>
            </span>
            <span
              class="c14"
            >
              Add new folder
            </span>
          </button>
          <button
            aria-disabled="false"
            class="c11 c12 c15"
          >
            <span
              aria-hidden="true"
              class="c8"
            >
              <svg
                fill="currentColor"
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M28.5 16a1.5 1.5 0 0 1-1.5 1.5h-9.5V27a1.5 1.5 0 1 1-3 0v-9.5H5a1.5 1.5 0 1 1 0-3h9.5V5a1.5 1.5 0 1 1 3 0v9.5H27a1.5 1.5 0 0 1 1.5 1.5"
                />
              </svg>
            </span>
            <span
              class="c14"
            >
              Add new assets
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <span
    class="c16"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
